src/networking.c

```diff
@@ -0,0 +1,374 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+/** \file
+ *
+ * Networking routines, used to setup networking (currently docker specific)
+ *
+ */
+
+#include <string.h>
+#include <stdbool.h>
+#include <signal.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <linux/if.h>
+#include <linux/if_tun.h>
+#include <ifaddrs.h>
+#include <arpa/inet.h>
+#include <net/if_arp.h>
+
+#include <glib.h>
+#include <glib/gprintf.h>
+
+#include "oci.h"
+#include "util.h"
+
+#define TUNDEV "/dev/net/tun"
+
+/*!
+ * Request to create a named tap interface
+ *
+ * \param tap \c tap interface name to create
+ *
+ * \return \c true on success, else \c false.
+ */
+static gboolean
+cc_oci_tap_create (const gchar *tap)
+{
+	struct ifreq ifr;
+	int fd;
+
+	if( (fd = open(TUNDEV, O_RDWR)) < 0 ) {
+		g_critical("Failed to open [%s] [%s]", TUNDEV, strerror(errno));
+		return false;
+	}
+
+	memset(&ifr, 0, sizeof(ifr));
+	ifr.ifr_flags = IFF_TAP;
+	g_strlcpy(ifr.ifr_name, tap, IFNAMSIZ);
+
+	if( ioctl(fd, TUNSETIFF, (void *) &ifr) < 0 ) {
+		g_critical("Failed to create tap [%s] [%s]",
+			   tap, strerror(errno));
+		close(fd);
+		return false;
+	}
+
+	if( ioctl(fd, TUNSETPERSIST, 1) < 0 ){
+		g_critical("failed to TUNSETPERSIST [%s] [%s]",
+			   tap, strerror(errno));
+		close(fd);
+		return false;
+	}
+
+	close(fd);
+	return true;
+}
+
+
+static gboolean
+cc_oci_netlink_run(const gchar *cmd_line) {
+	gint exit_status = -1;
+	gboolean ret = false;
+
+	ret = g_spawn_command_line_sync(cmd_line,
+				  NULL,
+				  NULL,
+				  &exit_status,
+				  NULL);
+
+	if ( !ret ) {
+		g_critical("failed to spawn [%s]", cmd_line);
+		return false;
+	}
+
+	if (exit_status) {
+		g_critical("netlink failure [%s] [%d]", cmd_line, exit_status);
+		return false;
+	}
+
+	return true;
+}
+
+/*!
+ * Request to create the networking framework
+ * that will be used to connect the specified
+ * container network (veth) to the VM
+ *
+ * The container may be associated with multiple
+ * networks and function has to be invoked for
+ * each of those networks
+ *
+ * Once the OCI spec supports the creation of 
+ * VM compatible tap interfaces in the network
+ * plugin, this setup will not be required
+ *
+ * \param netconfig \c networking configuration
+ *
+ * \return \c true on success, else \c false.
+ */
+gboolean
+cc_oci_network_create (const struct cc_oci_config * const config) {
+	gchar cmd_line[1024];
+
+	/* No networking enabled */
+	if (config->net.ifname == NULL) {
+		return true;
+	}
+
+	if (!cc_oci_tap_create(config->net.tap_device)) {
+		return false;
+	}
+
+	/* Place holder till we create these using appropriate
+	 * netlink commands
+	 */
+
+#define NETLINK_CREATE_BRIDGE "ip link add name %s type bridge"
```

Comment:
please use absolute paths, is safer an portable ([example](https://github.com/clearlinux/micro-config-drive/blob/master/configure.ac#L72))